crates/ty_python_semantic/src/types/call/bind.rs

```diff
@@ -2623,7 +2626,22 @@ impl<'a, 'db> ArgumentTypeChecker<'a, 'db> {
             .apply_specialization(self.db, isolated_specialization);
 
         let mut try_infer_tcx = || {
-            let return_ty = self.signature.return_ty?;
+            // For generic constructors, we use the type context to infer the specialization of the class
+            // instance instead of the method's return type.
+            let (inference_context, generic_constructor) = if let Type::BoundMethod(method) =
```

Comment:
Matching on explicit type variants is tempting, I know, but will break down when  the callable is not precisely a bound method. For example, in a case like the following, the callable would be a union type:
```py
class X[T]:
    def __init__(self, value: T):
        self.value = value

def _(flag: bool):
    b: X[int | None] = X(1) if flag else X(2)
    reveal_type(b)
```

This might be difficult to fix here, and might be an acceptable limitation for now, but we should maybe add a test case and a TODO?